pkg/vere/disk.c

```diff
@@ -867,66 +868,494 @@ u3_disk_init(c3_c* pax_c, u3_disk_cb cb_u)
+    //  running binary, then we need to create a new epoch
+    if ( 0 != strcmp(ver_c, URBIT_VERSION) ) {
+      if ( c3n == u3_disk_epoc_init(log_u) ) {
+        fprintf(stderr, "disk: failed to initialize epoch\r\n");
+        c3_free(log_u);
+        return 0;
+      }
+    }
+
+    //  mark the log as live
+    log_u->liv_o = c3y;
   }
 
-  log_u->liv_o = c3y;
 
 #if defined(DISK_TRACE_JAM) || defined(DISK_TRACE_CUE)
   u3t_trace_open(pax_c);
 #endif
 
   return log_u;
 }
+
+/* u3_disk_epoc_good: check for valid epoch.
+*/
+c3_o u3_disk_epoc_good(u3_disk* log_u, c3_d epo_d) {
+  /*  a "valid" epoch is currently defined as a writable folder in 
+   *  the <pier>/.urb/log directory named with a @ui and contains:
+   *  - a writable data.mdb file which can be opened by lmdb (add db checking?)
+   *  - a writable epoc.txt file
+   *  - a writable vere.txt file
+   *
+   *  XX: should we check if the correct event sequence exists in the data.mdb file?
+   *
+   *  note that this does not check if the epoch even contains snapshot files, 
+   *  let alone whether they're valid or not
+   */
+
+  c3_o ret_o = c3y;  //  return code
+
+  //  file paths
+  c3_c epo_c[8193], dat_c[8193], epv_c[8193], biv_c[8193];
+  snprintf(epo_c, sizeof(epo_c), "%s/0i%" PRIc3_d, log_u->com_u->pax_c, epo_d);
+  snprintf(dat_c, sizeof(dat_c), "%s/data.mdb", epo_c);
+  snprintf(epv_c, sizeof(epv_c), "%s/epoc.txt", epo_c);
+  snprintf(biv_c, sizeof(biv_c), "%s/vere.txt", epo_c);
+
+  c3_o dir_o = c3n;  //  directory is writable
+  c3_o dat_o = c3n;  //  data.mdb is writable
+  c3_o mdb_o = c3n;  //  data.mdb can be opened
+  c3_o epv_o = c3n;  //  epoc.txt is writable
+  c3_o biv_o = c3n;  //  vere.txt is writable
+
+  //  check if dir is writable
+  if ( 0 == access(epo_c, W_OK) ) {
+    dir_o = c3y;
+  }
+
+  //  check if data.mdb is writable
+  if ( 0 == access(dat_c, W_OK) ) {
+    dat_o = c3y;
+    //  check if we can open data.mdb
+    MDB_env* env_u;
+    if ( 0 != (env_u = u3_lmdb_init(epo_c)) ) {
+      mdb_o = c3y;
+    }
+    u3_lmdb_exit(env_u);
+  }
+
+  //  check if epoc.txt is writable
+  if ( 0 == access(epv_c, W_OK) ) {
+    epv_o = c3y;
+  }
+
+  //  check if vere.txt is writable
+  if ( 0 == access(biv_c, W_OK) ) {
+    biv_o = c3y;
+  }
+
+  //  print error messages
+  if ( c3n == dir_o ) {
+    fprintf(stderr, "disk: epoch 0i%" PRIc3_d " is not writable\r\n", epo_d);
+    ret_o = c3n;
+  }
+  if ( c3n == dat_o ) {
+    fprintf(stderr, "disk: epoch 0i%" PRIc3_d "/data.mdb is not writable\r\n", epo_d);
+    ret_o = c3n;
+  }
+  if ( c3n == mdb_o ) {
+    fprintf(stderr, "disk: epoch 0i%" PRIc3_d "/data.mdb can't be opened\r\n", epo_d);
+    ret_o = c3n;
+  }
+  if ( c3n == epv_o ) {
+    fprintf(stderr, "disk: epoch 0i%" PRIc3_d "/epoc.txt is not writable\r\n", epo_d);
+    ret_o = c3n;
+  }
+  if ( c3n == biv_o ) {
+    fprintf(stderr, "disk: epoch 0i%" PRIc3_d "/vere.txt is not writable\r\n", epo_d);
+    ret_o = c3n;
+  }
+
+  return ret_o;
+}
```

Comment:
This function makes me nervous. The validation that it performs seems sane, but it's then used to decide whether we should overwrite this epoc or not (in `u3_disk_epoc_init()`). It doesn't seem quite constrained enough for that purpose. I think we can only do that if there are no events in the log for the epoc under consideration.

Reply:
Good call-- I added a block that checks whether or not there are any events in the epoch's `data.mdb` file.